src/adapter/src/coord/command_handler.rs

```diff
@@ -888,13 +888,21 @@ impl Coordinator {
             // Acquire read holds _before_ we determine the least valid read.
             // Otherwise, we're not guaranteed that the since frontier doesn't
             // advance forward from underneath us.
+            let read_holds = self
+                .acquire_read_holds(timestamp, &ids, false)
+                .expect("precise==false, so acquiring read holds always succeeds");
+            let least_valid_read = self.least_valid_read(&read_holds);
+            timestamp.advance_by(least_valid_read.borrow());
+
+            // NOTE: The Drop impl of ReadHolds makes sure that the hold is
+            // released when we don't use it.
             if acquire_read_holds {
-                self.acquire_read_holds_auto_cleanup(session, timestamp, &ids, false)
-                    .expect("precise==false, so acquiring read holds always succeeds");
+                self.txn_read_holds
+                    .entry(session.conn_id().clone())
+                    .or_insert_with(Vec::new)
+                    .push(read_holds);
```

Comment:
I don't really remember when `acquire_read_holds` is `true` or `false`, but I'm a little scared that we're breaking some invariant here without realizing. Now we always get a read hold, but immediately release it if `acquire_read_holds` is `false`. Any idea what's going on here?

Reply:
Ok, I actually looked it up. We pass in `true` when we're actually creating a materialized view and `false` when we're running `EXPLAIN` for creating a materialized view. So this is probably fine? It's making `EXPLAIN` more expensive because we have to install and then release read holds.

Reply:
I did it so that there's only this one version of `least_valid_read`, but if we're concerned I could do custom code that only looks at the collection sinces, without acquiring read holds. But I had the same thought when writing it... :see_no_evil:

Also, though, I don't think we'll be hard-pressed to optimize throughput/latency of `EXPLAIN ...` queries anytime soon... :sweat_smile:

Reply:
Yeah agreed, I think it's fine as is.

Reply:
> Also, though, I don't think we'll be hard-pressed to optimize throughput/latency of EXPLAIN ... queries anytime soon... 😅

Yeah, I'd say this is fine.